portfolio/src/main/java/com/google/sps/servlets/FiltersServlet.java

```diff
@@ -0,0 +1,70 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Iterator;
+import com.google.gson.Gson;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.sps.data.Comment;
+import com.google.sps.data.AllComments;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+
+@WebServlet("/filters")
+public final class FiltersServlet extends HttpServlet {
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        // Get the filter input from the form.
+        String filter = getParameter(request, "filter-comments", null);
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query queryAllComments = new Query("AllComments");
+        PreparedQuery resultsAllComments = datastore.prepare(queryAllComments);
+        Iterator<Entity> iterAllComments = resultsAllComments.asIterator();
+        Entity entityAllComments = iterAllComments.next();
+
+        // Update the filter property
+        entityAllComments.setProperty("filter", filter);
+
+        // Add the updated entity back in the datastore
+        datastore.put(entityAllComments);
+
+        response.sendRedirect("/contact.html");
+        return;
+    }
+
+    /** Returns the desired parameter entered by the user, or null if the user input was invalid. */
+    private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+        // Get the input from the form.
+        String value = request.getParameter(name);
+
+        if (value == null) {
+            return defaultValue;
+        }
+
+        return value;
+    }
+}
```

Comment:
Anand suggested returning an Optional in getParameter so that if you have a null value you could do <optional_name>.orElse(<default_value>) if the parameter is null.
https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html

Reply:
The way I did it, I returned Optional.ofNullable(<value_of_parameter>) in getParameter where <value_of_parameter> is a String. Then, wherever I called getParameter, I did getParameter(request, parameterName).orElse(<default_value>). This automatically converts to a String. It might make more sense if you looked at my [code](https://github.com/kashiish/step/blob/0ac4eebbd2d947d3bbe1b5f42ce8e4f7e476e623/portfolio/src/main/java/com/google/sps/servlets/DataServlet.java#L129). I think it's related to readability because it makes it clear the orElse value is a default if the parameter value is invalid. I hope that made sense!

Reply:
Thank you! I will look into that then.